resharper/resharper-unity/src/HlslSupport/Integration/Injections/InjectedHlslFileLocationTracker.cs

```diff
@@ -68,91 +69,168 @@ private IEnumerable<CppFileLocation> GetIncludesLocation(IPsiSourceFile sourceFi
         {
             // PSI is not commited here
             // TODO: cpp global cache should calculate cache only when PSI for file with cpp injects is committed.
             
             var sourceFile = cppFileLocation.GetRandomSourceFile(mySolution);
             var range = cppFileLocation.RootRange;
             Assertion.Assert(range.IsValid, "range.IsValid");
-
+            
             var buffer = sourceFile.Document.Buffer;
             var type = GetShaderProgramType(buffer, range.StartOffset);
-            var includeType = GetIncludeProgramType(type);
 
-            yield return new CppFileLocation(myCppExternalModule, mySolution.SolutionDirectory.Combine(Utils.ShaderConfigFile));
-            
-            if (includeType != InjectedHlslProgramType.Uknown)
+
+            var defines = GetDefinedMacroses(ProjectedBuffer.Create(buffer, range), out var isSurface);
+            var includes = GetIncludes(sourceFile, type, isSurface);
+
+            return (includes, defines);
+        }
+
+        private Dictionary<string, string> GetDefinedMacroses(IBuffer buffer,out bool isSurface)
+        {
+            isSurface = false;
+            var lexer = CppLexer.Create(buffer);
+            lexer.Start();
+
+            var definedMacroses = new Dictionary<string, string>();
+            var shaderTarget = "25";
+            while (lexer.TokenType != null)
             {
-                var includes = GetIncludesLocation(sourceFile, includeType);
-                foreach (var include in includes)
+                var tokenType = lexer.TokenType;
+                if (tokenType is CppDirectiveTokenNodeType)
                 {
-                    yield return include;
+                    lexer.Advance();
+                    var context = lexer.GetTokenText().TrimStart();
+                    var (pragmaType, firstValue) = GetPragmaAndValue(context);
+                    if (pragmaType.Equals("surface"))
+                        isSurface = true;
+
+                    // based on https://docs.unity3d.com/Manual/SL-ShaderPrograms.html
+                    // We do not have any solution how we could handle multi_compile direcitves. It is complex task because
+                    // a lot of different files could be produces from multi_compile combination
+                    // Right now, we will consider first combination.
+                    if (!firstValue.IsEmpty() && (pragmaType.Equals("multi_compile") || pragmaType.Equals("multi_compile_local") ||
+                                                  pragmaType.Equals("shader_feature_local") || pragmaType.Equals("shader_feature")))
+                    {
+                        definedMacroses[firstValue] = "1";
+                    }
+
+                    if (pragmaType.Equals("target"))
+                        shaderTarget = firstValue.Replace(".", "");
```

Comment:
It's not obvious that `shaderTarget` is a shader version. Perhaps replace the "25", "40" and "50" magic strings with constants that explain the name? Also, is it legal to specify both `target` and `geometry` or `hull`? What should the target version be then?

Reply:
```
            #pragma target 3.0
            #pragma geometry geom
 
            #if SHADER_TARGET == 40
                #error GEOMETRY
            #endif

```
I got "GEOMETRY" error in Unity when compiling that shader. That means that geometry is preferred by Unity thatn target directive

Reply:
Target is maximum version then, I replaced literals by constants with names